mbedtls: make `PSA_WANT_KEY_TYPE_xxx_KEY_PAIR_BASIC` promptless

valeriosetti opened the pull request:
`PSA_WANT_KEY_TYPE_xxx_KEY_PAIR_BASIC` will be selected by any of:

- `PSA_WANT_KEY_TYPE_xxx_KEY_PAIR_IMPORT`
- `PSA_WANT_KEY_TYPE_xxx_KEY_PAIR_EXPORT`
- `PSA_WANT_KEY_TYPE_xxx_KEY_PAIR_GENERATE`
- `PSA_WANT_KEY_TYPE_xxx_KEY_PAIR_DERIVE`. 

This mechanism mimic what Mbed TLS already does internally with the corresponding build symbols.

valeriosetti committed:
mbedtls: auto-enable PSA_WANT_KEY_TYPE_xxx_KEY_PAIR_BASIC

PSA_WANT_KEY_TYPE_xxx_KEY_PAIR_BASIC build symbols are automatically
enabled in Mbed TLS header files whenever any key pair feature between
IMPORT,EXPORT, GENERATE,DERIVE is set. So we mimic the same behavior with
Kconfig symbols:
- do not add BASIC to the automatic generated Kconfig file;
- let BASIC be auto-enabled as soon as any other feature (IMPORT,EXPORT,
  GENERATE,DERIVE) is enabled for the same key type.

The 2nd point is achieved by adding a new Kconfig file which is meant
to hold the logic between PSA_WANT symbols. This is necessary because
Kconfig.psa is automatically generated.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
mbedtls: rename automatically generated Kconfig file for PSA symbols

Rename Kconfig.psa to Kconfig.psa.auto to emphasize that this file
is automatically generated.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font commented on modules/mbedtls/Kconfig.psa:
```diff
@@ -301,7 +301,7 @@ config PSA_WANT_KEY_TYPE_RSA_PUBLIC_KEY
 	default y if PSA_CRYPTO_ENABLE_ALL
 
 config PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_BASIC
-	bool "PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_BASIC" if !MBEDTLS_PROMPTLESS
+	bool
 	default y if PSA_CRYPTO_ENABLE_ALL
```
Remove those `PSA_WANT_KEY_TYPE_*_KEY_PAIR_BASIC` completely from this file as you are adding them to `Kconfig.psa.logic`?

tomi-font commented on modules/mbedtls/Kconfig.psa.logic:
```diff
@@ -0,0 +1,24 @@
+# some logic between PSA_WANT symbols.
+
+config PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_BASIC
+	default y
+	depends on PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_IMPORT || \
+		   PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_EXPORT || \
+		   PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_GENERATE || \
+		   PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_DERIVE
```
In your commit message you say that the features select `*_BASIC`, but you are doing it the other way around in practice (`*_BASIC` auto-enable themselves).

tomi-font commented on modules/mbedtls/create_psa_files.py:
```diff
@@ -57,6 +57,20 @@
 
 H_FOOTER="\n#endif /* CONFIG_PSA_H */\n"
 
+# In Mbed TLS the PSA_WANT_KEY_TYPE_[ECC|RSA|DH]_KEY_PAIR_BASIC build symbols
+# are automatically enabled whenever any other _IMPORT, _EXPORT, _GENERATE or
+# _DERIVE feature is set for the same key type
+# (see "modules/crypto/mbedtls/include/psa/crypto_adjust_config_key_pair_types.h").
+# Therefore we mimic the same pattern with Kconfigs as follows:
+# - do not add _BASIC Kconfigs to the automatic generated file (KCONFIG_PATH);
+# - add _BASIC Kconfigs to Kconfig.psa.logic and let them "default y" as soon as
+#   any other _IMPORT, _EXPORT, _GENERATE or _DERIVE Kconfigs are enabled.
```
Not sure such a length comment was needed (could have been more generic), but it's good as is to me. :slightly_smiling_face:

valeriosetti committed:
tests: tls_configurations: adjust Kconfig in overlay-ec

CONFIG_PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_BASIC is promptless so it cannot
be selected. Moreover it's also automatically enabled by
CONFIG_PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_GENERATE in the same overlay file
so there would be no need to explicitly enable it.

As for the IMPORT, EXPORT, DERIVE they are needed for the TLS connection
to work properly. Previously it was working because at least IMPORT and
EXPORT are internally enabled by Mbed TLS at build time. So here we
are basically doing the same enablements with Kconfigs in clear.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

frkv reviewed (approved):
LGTM
